File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.swift
import Foundation

func solution(_ s:String) -> Bool
{   
    var count: Int = 0
    var stack: [Character] = Array(s)
    
    for char in stack {
        if char == "(" {
            // 여는 괄호면 count +1
            count += 1
        } else {
            // 닫는 괄호면 count -1
            count -= 1
            
            // 만약 count가 음수가 된다면 불가능한 괄호 쌍
            if count < 0 {
                break
            }
        }
    }

    return count == 0
}

File: 프로그래머스/2/12909. 올바른 괄호/올바른 괄호.js
function solution(s){
    var answer = true;
    const stack = []
    
    if (s[0] === ")") answer = false
    
    for (const char of s) {
        if (char === "(") {
            stack.push(char)
        } else {
            if (char[char.length-1] === ")") {
                stack.pop()
            } else {
                answer = false
                break;
            }
        }
    }
    
    if (stack.length !== 0 ) answer = false

    return answer;
}